File: frontend/src/services/holidayService.ts
import api from './api';

export interface HolidayMonth {
  hospital_id: string;
  year: number;
  month: number;
  holiday_days: number[];
  festival_days: number[];
  working_days: number;
  total_days: number;
  created_at: string | null;
  updated_at: string | null;
}

const holidayService = {
  getMonth: (year: number, month: number) =>
    api.get<HolidayMonth>(`/holidays/${year}/${month}`).then(res => res.data),

  getYear: (year: number) =>
    api.get<HolidayMonth[]>(`/holidays/${year}`).then(res => res.data),

  saveMonth: (year: number, month: number, holiday_days: number[], festival_days: number[]) =>
    api.put<HolidayMonth>(`/holidays/${year}/${month}`, { holiday_days, festival_days }).then(res => res.data),
};

export default holidayService;

File: backend/app/services/holiday_service.py
"""
Holiday allocation service — one row per hospital per month, storing which
day-of-month numbers are marked as holidays.
"""
import uuid
from sqlalchemy.orm import Session
from ..models.attendance import Holidays


def get_month(db: Session, hospital_id: uuid.UUID, year: int, month: int) -> Holidays | None:
    return (
        db.query(Holidays)
        .filter(Holidays.hospital_id == hospital_id, Holidays.year == year, Holidays.month == month)
        .first()
    )


def get_year(db: Session, hospital_id: uuid.UUID, year: int) -> list[Holidays]:
    return (
        db.query(Holidays)
        .filter(Holidays.hospital_id == hospital_id, Holidays.year == year)
        .order_by(Holidays.month)
        .all()
    )


def upsert_month(
    db: Session, hospital_id: uuid.UUID, year: int, month: int,
    holiday_days: list[int], festival_days: list[int], updated_by: uuid.UUID,
) -> Holidays:
    row = get_month(db, hospital_id, year, month)
    if row:
        row.holiday_days = holiday_days
        row.festival_days = festival_days
    else:
        row = Holidays(
            hospital_id=hospital_id, year=year, month=month,
            holiday_days=holiday_days, festival_days=festival_days, created_by=updated_by,
        )
        db.add(row)
    db.commit()
    db.refresh(row)
    return row
